armsrc/legicrf.c

```diff
@@ -40,6 +43,458 @@ static int      legic_reqresp_drift;
+}
+
+// Returns a demedulated bit
+//
+// An aproximated power measurement is available every 18.9us. The bit time
+// is 100us. The code samples 5 times and uses the last (most stable) sample.
+//
+// Note: The demodulator would be drifting (18.9us * 5 != 100us), rx_frame_as_reader
+// has a delay loop that aligns rx_bit_as_reader calls to the TAG tx timeslots.
+static inline bool rx_bit_as_reader() {
+  int32_t power;
+
+  for(size_t i = 0; i<5; ++i) {
+    power = sample_power();
+  }
+
+  return (power > INPUT_THRESHOLD);
+}
+
+//-----------------------------------------------------------------------------
+// Modulation (Reader)
+//
+// I've tried to modulate the Legic specific pause-puls using ssc and the default
+// ssc clock of 105.4 kHz (bit periode of 9.4us) - previous commit. However,
+// the timing was not precise enough. By increasing the ssc clock this could
+// be circumvented, but the adventage over bitbang would be little.
+//-----------------------------------------------------------------------------
+
+static inline void tx_bit_as_reader(bool bit) {
+  // insert pause
+  LOW(GPIO_SSC_DOUT);
+  last_frame_end += RWD_TIME_PAUSE;
+  while(GET_TICKS < last_frame_end) { };
+  HIGH(GPIO_SSC_DOUT);
+
+  // return to high, wait for bit periode to end
+  last_frame_end += (bit ? RWD_TIME_1 : RWD_TIME_0) - RWD_TIME_PAUSE;
+  while(GET_TICKS < last_frame_end) { };
+}
+
+//-----------------------------------------------------------------------------
+// Frame Handling (Reader)
+//
+// The LEGIC RF protocol from card to reader does not include explicit frame
+// start/stop information or length information. The reader must know beforehand
+// how many bits it wants to receive.
+// Notably: a card sending a stream of 0-bits is indistinguishable from no card
+// present.
+//-----------------------------------------------------------------------------
+
+static void tx_frame_as_reader(uint32_t frame, uint8_t len) {
+  FpgaWriteConfWord(FPGA_MAJOR_MODE_HF_READER_TX);
+
+  // wait for next tx timeslot
+  last_frame_end += RWD_FRAME_WAIT;
+  while(GET_TICKS < last_frame_end) { };
+
+  // transmit frame, MSB first
+  for(uint8_t i = 0; i < len; ++i) {
+    bool bit = (frame >> i) & 0x01;
+    tx_bit_as_reader(bit ^ legic_prng_get_bit());
+    legic_prng_forward(1);
+  };
+
+  // add pause to mark end of the frame
+  LOW(GPIO_SSC_DOUT);
+  last_frame_end += RWD_TIME_PAUSE;
+  while(GET_TICKS < last_frame_end) { };
+  HIGH(GPIO_SSC_DOUT);
+}
+
+static uint32_t rx_frame_as_reader(uint8_t len) {
+  FpgaWriteConfWord(FPGA_MAJOR_MODE_HF_READER_RX_XCORR
+                  | FPGA_HF_READER_RX_XCORR_848_KHZ
+                  | FPGA_HF_READER_RX_XCORR_QUARTER_FREQ);
+
+  // hold sampling until card is expected to respond
+  last_frame_end += TAG_FRAME_WAIT;
+  while(GET_TICKS < last_frame_end) { };
+
+  uint32_t frame = 0;
+  for(uint8_t i = 0; i < len; i++) {
+    frame |= (rx_bit_as_reader() ^ legic_prng_get_bit()) << i;
```

Comment:
Need to check if there is enough time for `legic_prng_get_bit()` and `legic_prng_forward()` with the new xcorr data rates. It may be necessary to precalculate the keystream (as it was before).

Reply:
Thanks for the hint, I think with the absolute timing, I added, we can do almost anything in inter-bit gabs without loosing sync. As long as timing permits I try to put code cleanness over performance.

Reply:
I'ts not about being synchronous with SSP but to be precise on the bit timing of Legic. The tag does a simple on/off keying with a bit time of 100us and a delay of exactly 330us after the last reader pulse.

As discussed above the SSP time is a little too fast. So eventually the SSP clock will have drifted ~10us thats after ~2 bits. Now the absolute wait implicitly skips 1 Q or I (we don't rely know) and the sampling is re-adjusted.

Reply:
All zeroes is equal to no signal at all, so there is no guaranteed edge to synchronise to.

Reply:
> As discussed above the SSP time is a little too fast. So eventually the SSP clock will have drifted ~10us thats after ~2 bits. Now the absolute wait implicitly skips 1 Q or I (we don't rely know) and the sampling is re-adjusted.

Yep, true for the current sampling rate. Will no longer be necessary with the quadruple sample rate (21 I/Q pairs @ 212kHz = 99,1us)

Reply:
Why go for 99% accuracy when there is a 99.9% one?. In a write that tales up to 3.6ms the drift is 36% of the bit periode.

Reply:
99.1us is 100% accuracy. 100us is an approximation.

Reply:
mea culpa

Reply:
What you think of speeding up SSP in `FPGA_MAJOR_MODE_HF_READER_TX`to match 212kH and use it trough the code to run the prng?

Reply:
Not sure if this would add some benefit?

Reply:
`StartCountSspClk` and `GetCountSspClk` would count in 4.7us increments. Giving as a perfect time source.

Without the suggested change the clock changes between RX and TX making SspClk almost unusable.

Reply:
Ah, I see, you are already a step ahead and think about Legic simulation. Would it be an issue if the clock is faster (but both TX and RX the same)?

Reply:
Don't panic! :smile:
I don't plan to increase the sampling rate or the frame rate. We would still get one Q/I pair each  4.7us. But with an increased SSP clock we could
* increase the frame size even further to 32bits, which would eliminate the need to clip the Q/I values
* and/or allow to finish the transfer from FPGA to ARM in a shorter time.

Reply:
But anyway I will wait until this PR is completed...

Reply:
No need to wait, I'm currently trying to get my hands on a real reader so I can extract the perfect pause-pulse modulation. Next to that the code already works.

Reply:
If it works, then please raise a PR. It will be better than the existing code and I don't want us to break each others code (btw: I just ordered an ISO15693 reader - the hf 15 sim and record commands code CANNOT WORK in my opinion).